python/across_api/burstcube/schema.py

```diff
@@ -0,0 +1,219 @@
+    """
+    Schema describing a BurstCube TOO Request.
+    """
+
+    id: Optional[str] = None
+    created_by: str
+    created_on: AstropyTime
+    modified_by: Optional[str] = None
+    modified_on: Optional[AstropyTime] = None
+    trigger_time: AstropyTime
+    trigger_info: BurstCubeTriggerInfo
+    exposure: AstropySeconds
+    offset: AstropySeconds
+    reject_reason: TOOReason = TOOReason.none
+    status: TOOStatus = TOOStatus.requested
+    too_info: str = ""
+
+
+class BurstCubeTOODelSchema(BaseSchema):
+    """
+    Schema for BurstCubeTOO DELETE API call.
+
+    Attributes
+    ----------
+    id
+        The ID of the BurstCubeTOODel object.
+    """
+
+    id: str
+
+
+class BurstCubeTOOPostSchema(OptionalPositionSchema):
+    """
+    Schema to submit a TOO request for BurstCube.
+    """
+
+    trigger_time: AstropyTime
+    trigger_info: BurstCubeTriggerInfo
+    exposure: AstropySeconds = 200 * u.s
+    offset: AstropySeconds = -50 * u.s
+
+
+class BurstCubeTOOGetSchema(BaseSchema):
+    """
+    Schema for BurstCubeTOO GET request.
+    """
+
+    id: str
+
+
+class BurstCubeTOOPutSchema(BurstCubeTOOPostSchema):
+    """
+    Schema for BurstCubeTOO PUT request.
+    """
+
+    id: str
+
+
+class BurstCubeTOORequestsGetSchema(OptionalDateRangeSchema):
+    """
+    Schema for GET requests to retrieve BurstCube Target of Opportunity (TOO) requests.
+    """
+
+    length: Optional[u.Quantity] = None
+    limit: Optional[int] = None
+
+    @model_validator(mode="after")
+    def check_begin_and_end_or_length_set(self):
+        if self.begin is not None and self.end is not None and self.length is not None:
+            raise ValueError("Cannot set both begin and end and length.")
+        elif self.begin is not None and self.length is not None:
+            self.end = self.begin + self.length
+        elif self.begin is None and self.end is None and self.length is not None:
+            self.end = Time.now()
+            self.begin = self.end - self.length
+
+
+class BurstCubeTOORequestsSchema(BaseSchema):
+    """
+    Schema for BurstCube TOO requests.
+    """
+
+    entries: List[BurstCubeTOOSchema]
+
+
+class BurstCubeTOOModel(BaseSchema):
+    """Database Model Schema for BurstCube TOO requests."""
+
+    __tablename__ = "burstcube_too"
+    ra: Optional[Decimal] = None
+    dec: Optional[Decimal] = None
+    error_radius: Optional[Decimal] = None
+    created_by: str
+    created_on: str
+    modified_by: Optional[str] = None
+    modified_on: Optional[str] = None
+    trigger_time: str
+    trigger_info: JsonStr
+    exposure: Decimal = Decimal(200)
+    offset: Decimal = Decimal(-50)
+    reject_reason: str = "None"
+    status: str = "Requested"
+    too_info: str = ""
+
+    @computed_field  # type: ignore
+    @property
+    def id(self) -> str:
+        return hashlib.md5(
+            f"{self.trigger_time}{self.exposure}{self.offset}".encode()
+        ).hexdigest()
```

Comment:
Remind me again, why do models need to be hashable?

Reply:
This code is just creating a unique ID key for each TOO based on the three unique required parameters (and no more).

Reply:
Ah, so the reason is that you decided to use these to form the partition key for this table?